packages/cli/src/commands/hydrogen/dev.ts

```diff
@@ -69,25 +69,47 @@ async function runDev({
     return [fileRelative, path.resolve(root, fileRelative)] as const;
   };
 
+  const serverBundleExists = () => file.exists(buildPathWorkerFile);
+
+  let miniOxygenStarted = false;
+  async function safeStartMiniOxygen() {
+    if (miniOxygenStarted) return;
+
+    await startMiniOxygen({
+      root,
+      port,
+      watch: true,
+      buildPathWorkerFile,
+      buildPathClient,
+    });
+
+    miniOxygenStarted = true;
+
+    const showUpgrade = await checkingHydrogenVersion;
+    if (showUpgrade) showUpgrade();
+  }
+
   const {watch} = await import('@remix-run/dev/dist/compiler/watch.js');
   await watch(await reloadConfig(), {
     reloadConfig,
     mode: process.env.NODE_ENV as any,
     async onInitialBuild() {
       await copyingFiles;
 
-      console.timeEnd(LOG_INITIAL_BUILD);
-
-      await startMiniOxygen({
-        root,
-        port,
-        watch: true,
-        buildPathWorkerFile,
-        buildPathClient,
-      });
+      if (!(await serverBundleExists())) {
+        const {renderFatalError} = await import('@shopify/cli-kit/node/ui');
+        return renderFatalError({
+          name: 'BuildError',
+          type: 0,
+          message:
+            'MiniOxygen cannot start because the server bundle has not been generated.',
```

Comment:
I'm often concerned that folks won't know what MiniOxygen is. @gfscott do we ever expose this name anywhere? I have been referring this as `local Oxygen runtime` instead.

Reply:
We've been showing "MiniOxygen" in the dev command since the beginning, though:

<img width="445" alt="image" src="https://user-images.githubusercontent.com/1634092/219334469-aa491139-f421-4b43-a2fa-17d0b7500149.png">

Not against hiding it but if we do, we should be consistent and remove it there as well I guess?